app/scripts/controllers/network/provider-api-tests/shared-tests.js

```diff
@@ -143,65 +143,104 @@ export function testsForProviderType(providerType) {
 
     describe('methods that have a param to specify the block', () => {
       const supportingBlockParam = [
-        { name: 'eth_call', blockParamIndex: 1 },
-        { name: 'eth_getBalance', blockParamIndex: 1 },
-        { name: 'eth_getBlockByNumber', blockParamIndex: 0 },
-        { name: 'eth_getCode', blockParamIndex: 1 },
-        { name: 'eth_getStorageAt', blockParamIndex: 2 },
-        { name: 'eth_getTransactionCount', blockParamIndex: 1 },
+        {
+          name: 'eth_call',
+          blockParamIndex: 1,
+          numberOfParameters: 2,
+        },
+        {
+          name: 'eth_getBalance',
+          blockParamIndex: 1,
+          numberOfParameters: 2,
+        },
+        {
+          name: 'eth_getBlockByNumber',
+          blockParamIndex: 0,
+          numberOfParameters: 2,
+        },
+        { name: 'eth_getCode', blockParamIndex: 1, numberOfParameters: 2 },
+        {
+          name: 'eth_getStorageAt',
+          blockParamIndex: 2,
+          numberOfParameters: 3,
+        },
+        {
+          name: 'eth_getTransactionCount',
+          blockParamIndex: 1,
+          numberOfParameters: 2,
+        },
       ];
-      supportingBlockParam.forEach(({ name, blockParamIndex }) => {
-        describe(`method name: ${name}`, () => {
-          testsForRpcMethodSupportingBlockParam(name, {
-            providerType,
-            blockParamIndex,
+      supportingBlockParam.forEach(
+        ({ name, blockParamIndex, numberOfParameters }) => {
+          describe(`method name: ${name}`, () => {
+            testsForRpcMethodSupportingBlockParam(name, {
+              providerType,
+              blockParamIndex,
+              numberOfParameters,
+            });
           });
-        });
-      });
+        },
+      );
     });
 
     describe('methods that assume there is no block param', () => {
```

Comment:
I know this isn't the target of this PR so this is more of just an observation/question: this seems like an odd way to categorize the methods for testing?

Reply:
These groups are meant to reflect the caching strategies used in our middleware: either we ignore the method, or we cache it by the current block, or we cache it by the referenced block.

This "assume there is no block param" group is for methods that we cache by block, where the middleware doesn't bother looking for any block reference in the method parameters.

Reply:
Definitely lots of room for improvement with how these are organized. We can describe these categories better, and deduplicate the common behavior more effectively.